pkg/boot/survey/gradual.go

```diff
@@ -0,0 +1,110 @@
+package survey
+
+import (
+	"context"
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// GradualSurveyor queries a progressively larger subset of the
+// cluster keyspace.  It is recommended for bootstrapping large
+// clusters where multicast storms may degrade the service.
+//
+// See specs/survey.md for details.
+type GradualSurveyor struct {
+	// Factor controls how long the GradualSurveyor will wait for
+	// peers at each distance. The raw wait time is calculated by
+	// evaluating 'distance * factor', before applying min, max &
+	// jitter.  Defaults to 1.5.
+	Factor float64
+
+	// Min and Max wait durations.  Defaults to 5s and 90s.
+	Min, Max time.Duration
+
+	// If true, use the absolute wait durations calculated via min,
+	// max and factor.  By default, the final wait time is sampled
+	// uniformly from the interval (w/2, w), where 'w' is the wait
+	// duration.
+	DisableJitter bool
+	*Surveyor
+}
+
+func (g GradualSurveyor) FindPeers(ctx context.Context, ns string, opt ...discovery.Option) (<-chan peer.AddrInfo, error) {
+	ctxSurv, cancel := context.WithCancel(ctx)
+
+	found, err := g.Surveyor.FindPeers(ctxSurv, ns, append(opt, WithDistance(uint8(0)))...)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	out := make(chan peer.AddrInfo, 1)
+	go func() {
+		defer close(out)
+		defer cancel()
+
+		b := backoff.Backoff{
+			Factor: g.factor(),
+			Min:    g.min(),
+			Max:    g.max(),
+			Jitter: !g.DisableJitter,
+		}
+
+		for {
+			select {
+			case <-time.After(b.Duration()):
+				cancel()
+				ctxSurv, cancel = context.WithCancel(ctx)
+				defer cancel()
+
+				found, err = g.Surveyor.FindPeers(ctxSurv, ns, append(opt, WithDistance(uint8(b.Attempt())))...)
+				if err != nil {
+					g.log.WithError(err).Debug("retry failed")
+				}
+				continue
+
+			case info := <-found:
+				out <- info
+
+			case <-ctx.Done():
+				return
+			}
+			break
```

Comment:
I assume the `break` intended to abort the discovery process after an initial peer has been found?  I think we should simplify the logic and remove it.  I realize that users could get sloppy and fail to close the context, but I think that's okay because (1) this is consistent with the `FindPeers` API across all libp2p and (2) the ability to keep the `out` channel open is actually very useful.  A common pattern is to keep a long-running discovery channel and use it to periodically refresh peers.  We should support that pattern.

Reply:
Note:  the `continue` on line 67 can be removed as well.